sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```
@@ -1067,90 +1083,115 @@ public Object call(SQLDatabase db) throws Exception{
+                                    logger.log(Level.SEVERE, "There was a problem adding the " +
+                                                    "attachment "
+                                                    + usa + "to the datastore for document " + rev,
+                                            e);
+                                    throw e;
                                 }
                             }
-                        } else {
+                        }
+                    } else {
 
-                            try {
-                                if (preparedAttachments != null) {
-                                    for (String[] key : preparedAttachments.keySet()) {
-                                        String id = key[0];
-                                        String rev = key[1];
+                        try {
+                            if (preparedAttachments != null) {
+                                for (String[] key : preparedAttachments.keySet()) {
+                                    String id = key[0];
+                                    String rev = key[1];
+                                    try {
                                         BasicDocumentRevision doc = getDocumentInQueue(db, id, rev);
-                                        if(doc != null) {
-                                            for (PreparedAttachment att : preparedAttachments.get(key)) {
+                                        if (doc != null) {
+                                            for (PreparedAttachment att : preparedAttachments.get
+                                                    (key)) {
                                                 attachmentManager.addAttachment(db, att, doc);
                                             }
                                         }
+                                    } catch (DocumentNotFoundException e){
+                                        continue;
```

From my understanding it was expected that `getDocumentInQueue` could return null for the document at the specified revision and it was safe to continue. Since now  `getDocumentInQueue` throws `DocumentNotFoundException` instead of returning null it's fine just to continue to the loop. I'll add in a comment to explain this.